### ISSUE TITLE
compilation fails under macOS 10.13 with XCode 9

### ISSUE THREAD

**Issue by pmalhaire** [osxfuse/osxfuse#416]
[EDIT 27 09 2017]
the only remaining bug to build under macOS 10.13 concerns gen_bridge_metadata. 
the bug is in the os but there is a workaround thanks to @swt2c see [link](https://github.com/osxfuse/osxfuse/issues/416#issuecomment-330980225)
[END EDIT]

High Sierra will be released very soon and the project is not compiling using it.

Here is what I got when compiling osxfuse under High Sierra Beta with Xcode 9 (see after for how a drafted fix).

```bash 
CompileC build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/fuse_kext.o load_osxfuse/fuse_kext.c normal x86_64 c com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/pierrot/osxfuse/support
    export LANG=en_US.US-ASCII
    /Applications/XcodeBeta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c -arch x86_64 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -std=gnu99 -fmodules -gmodules -fmodules-prune-interval=86400 -fmodules-prune-after=345600 -fbuild-session-file=/var/folders/gx/g_xkk19x7_31_hc7nv4hlc400000gn/C/org.llvm.clang/ModuleCache/Session.modulevalidation -fmodules-validate-once-per-build-session -Wnon-modular-include-in-framework-module -Werror=non-modular-include-in-framework-module -Wno-trigraphs -fpascal-strings -Os -fno-common -Werror -Wno-missing-field-initializers -Wno-missing-prototypes -Werror=return-type -Wunreachable-code -Werror=deprecated-objc-isa-usage -Werror=objc-root-class -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wconditional-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -DOSXFUSE_ENABLE_MACFUSE_MODE=0 -isysroot /Applications/XcodeBeta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -mmacosx-version-min=10.13 -g -fvisibility=hidden -Wno-sign-conversion -Wno-infinite-recursion -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -iquote /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-generated-files.hmap -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-own-target-headers.hmap -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-all-target-headers.hmap -iquote /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-project-headers.hmap -I/tmp/osxfuse/fsbundle/include -I/Users/pierrot/osxfuse/support/../common -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/DerivedSources/x86_64 -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/DerivedSources -F/tmp/osxfuse/fsbundle -MMD -MT dependencies -MF /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/fuse_kext.d --serialize-diagnostics /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/fuse_kext.dia -c /Users/pierrot/osxfuse/support/load_osxfuse/fuse_kext.c -o /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/fuse_kext.o
/Users/pierrot/osxfuse/support/load_osxfuse/fuse_kext.c:322:27: error: use of undeclared identifier 'kOSKextReturnSystemPolicy'
        } else if (ret == kOSKextReturnSystemPolicy) {
                          ^
1 error generated.

CompileC build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/main.o load_osxfuse/main.c normal x86_64 c com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/pierrot/osxfuse/support
    export LANG=en_US.US-ASCII
    /Applications/XcodeBeta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c -arch x86_64 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -std=gnu99 -fmodules -gmodules -fmodules-prune-interval=86400 -fmodules-prune-after=345600 -fbuild-session-file=/var/folders/gx/g_xkk19x7_31_hc7nv4hlc400000gn/C/org.llvm.clang/ModuleCache/Session.modulevalidation -fmodules-validate-once-per-build-session -Wnon-modular-include-in-framework-module -Werror=non-modular-include-in-framework-module -Wno-trigraphs -fpascal-strings -Os -fno-common -Werror -Wno-missing-field-initializers -Wno-missing-prototypes -Werror=return-type -Wunreachable-code -Werror=deprecated-objc-isa-usage -Werror=objc-root-class -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wconditional-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wshorten-64-to-32 -Wpointer-sign -Wno-newline-eof -DOSXFUSE_ENABLE_MACFUSE_MODE=0 -isysroot /Applications/XcodeBeta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -mmacosx-version-min=10.13 -g -fvisibility=hidden -Wno-sign-conversion -Wno-infinite-recursion -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -iquote /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-generated-files.hmap -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-own-target-headers.hmap -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-all-target-headers.hmap -iquote /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/load_osxfuse-project-headers.hmap -I/tmp/osxfuse/fsbundle/include -I/Users/pierrot/osxfuse/support/../common -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/DerivedSources/x86_64 -I/Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/DerivedSources -F/tmp/osxfuse/fsbundle -MMD -MT dependencies -MF /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/main.d --serialize-diagnostics /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/main.dia -c /Users/pierrot/osxfuse/support/load_osxfuse/main.c -o /Users/pierrot/osxfuse/support/build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/main.o

** BUILD FAILED **


The following build commands failed:
	CompileC build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/i386/fuse_kext.o load_osxfuse/fuse_kext.c normal i386 c com.apple.compilers.llvm.clang.1_0.compiler
	CompileC build/load_osxfuse.build/Release/load_osxfuse.build/Objects-normal/x86_64/fuse_kext.o load_osxfuse/fuse_kext.c normal x86_64 c com.apple.compilers.llvm.clang.1_0.compiler
(2 failures)
T:fsbundle           | Failed to build target
[1]    27759 broken pipe  ./build.sh -t distribution -v 20 2>&1 |
       27760 terminated   tee /tmp/log
>>> elapsed time 13s
```

#### In support submodule :

```diff
diff --git a/Packages/Distribution/Plugins/PostInstall.xcodeproj/project.pbxproj b/Packages/Distribution/Plugins/PostInstall.xcodeproj/project.pbxproj
index 14821f0..f5e7818 100644
--- a/Packages/Distribution/Plugins/PostInstall.xcodeproj/project.pbxproj
+++ b/Packages/Distribution/Plugins/PostInstall.xcodeproj/project.pbxproj
@@ -296,8 +296,9 @@
                                INSTALL_PATH = "$(HOME)/Library/Bundles";
                                PRODUCT_BUNDLE_IDENTIFIER = com.github.osxfuse.installerplugin.PostInstall;
                                PRODUCT_NAME = "$(TARGET_NAME)";
+                               "VALID_ARCHS[sdk=macosx10.12]" = x86_64;
+                               "VALID_ARCHS[sdk=macosx10.13]" = x86_64;
                                "VALID_ARCHS[sdk=macosx10.5]" = "i386 ppc";
-                               "VALID_ARCHS[sdk=macosx10.12]" = "x86_64";
                                WRAPPER_EXTENSION = bundle;
                        };
                        name = Debug;
@@ -310,8 +311,9 @@
                                INSTALL_PATH = "$(HOME)/Library/Bundles";
                                PRODUCT_BUNDLE_IDENTIFIER = com.github.osxfuse.installerplugin.PostInstall;
                                PRODUCT_NAME = "$(TARGET_NAME)";
+                               "VALID_ARCHS[sdk=macosx10.12]" = x86_64;
+                               "VALID_ARCHS[sdk=macosx10.13]" = x86_64;
                                "VALID_ARCHS[sdk=macosx10.5]" = "i386 ppc";
-                               "VALID_ARCHS[sdk=macosx10.12]" = "x86_64";
                                WRAPPER_EXTENSION = bundle;
                        };
                        name = Release;
diff --git a/load_osxfuse/fuse_kext.c b/load_osxfuse/fuse_kext.c
index 2ee9815..d6bcb62 100644
--- a/load_osxfuse/fuse_kext.c
+++ b/load_osxfuse/fuse_kext.c
@@ -111,7 +111,7 @@
     ) __attribute__((weak_import));
 #endif

-#if MAC_OS_X_VERSION_MAX_ALLOWED < 101300
+#if MAC_OS_X_VERSION_MAX_ALLOWED < 101400
     #define kOSKextReturnSystemPolicy         -603946981
 #endif
```

#### in framework submodule : 

I just skipped the bridge support an ugly way just to make it work.

I replaced 

```bash
if [[ ! -x \"${GEN_BRIDGE_METADATA}\" ]]
```

with 
```bash
if [[ 0 ]]
```

entire patch

```diff
diff --git a/OSXFUSE.xcodeproj/project.pbxproj b/OSXFUSE.xcodeproj/project.pbxproj
index 6aef2d8..1854a62 100644
--- a/OSXFUSE.xcodeproj/project.pbxproj
+++ b/OSXFUSE.xcodeproj/project.pbxproj
@@ -201,7 +201,7 @@
 			);
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "if [[ \"${ACTION}\" != \"build\" ]]\nthen\n    exit 0\nfi\n\nGEN_BRIDGE_METADATA=\"/usr/bin/gen_bridge_metadata\"\nif [[ ! -x \"${GEN_BRIDGE_METADATA}\" ]]\nthen\n    echo \"Skipping BridgeSupport due to missing gen_bridge_metadata.\"\n    exit 0\nfi\n\nOUT_FRAMEWORK=\"${TARGET_BUILD_DIR}/${WRAPPER_NAME}\"\nBRIDGE_DIR=\"${TARGET_BUILD_DIR}/${UNLOCALIZED_RESOURCES_FOLDER_PATH}/BridgeSupport/\"\n\n/bin/mkdir -p \"${BRIDGE_DIR}\"\nDYLD_LIBRARY_PATH=\"${OSXFUSE_LIBRARY_SEARCH_PATHS}\" \"${GEN_BRIDGE_METADATA}\" -f \"${OUT_FRAMEWORK}\" > \"${BRIDGE_DIR}/${PRODUCT_NAME}.bridgesupport\"";
+			shellScript = "if [[ \"${ACTION}\" != \"build\" ]]\nthen\n    exit 0\nfi\n\nGEN_BRIDGE_METADATA=\"/usr/bin/gen_bridge_metadata\"\nif [[ 0 ]]\nthen\n    echo \"Skipping BridgeSupport due to missing gen_bridge_metadata.\"\n    exit 0\nfi\n\nOUT_FRAMEWORK=\"${TARGET_BUILD_DIR}/${WRAPPER_NAME}\"\nBRIDGE_DIR=\"${TARGET_BUILD_DIR}/${UNLOCALIZED_RESOURCES_FOLDER_PATH}/BridgeSupport/\"\n\n/bin/mkdir -p \"${BRIDGE_DIR}\"\nDYLD_LIBRARY_PATH=\"${OSXFUSE_LIBRARY_SEARCH_PATHS}\" \"${GEN_BRIDGE_METADATA}\" -f \"${OUT_FRAMEWORK}\" > \"${BRIDGE_DIR}/${PRODUCT_NAME}.bridgesupport\"";
 			showEnvVarsInLog = 0;
 		};
 /* End PBXShellScriptBuildPhase section */
```

### I ended up with a build success : 

The bridge support patch is not acceptable as is.

The rest of the patches look ok, don't they ?

What is `BridgeSupport` where can I find some doc to fix it cleanly ?

### The patches as files if someone need them : 

[framework_patch.txt](https://github.com/osxfuse/osxfuse/files/1318167/framework_patch.txt)
[support_patch2.txt](https://github.com/osxfuse/osxfuse/files/1318168/support_patch2.txt)

**Comment by swt2c** [comment]
For the error about `kOSKextReturnSystemPolicy`, that identifier is defined in the 10.13 SDKs in `libkern/OSKextLib.h`, so I think the proper fix would be to move those three lines to after where `libkern/OSKextLib.h` gets included.

**Comment by pmalhaire** [comment]
@swt2c it makes sense but just switching the order does not work I must be missing something

**Comment by swt2c** [comment]
@pmalhaire what version of 10.13 beta are you using?

**Comment by pmalhaire** [comment]
@swt2c 
macOS 10.13 17A360a
Xcode 9.0 beta (9M136h)

**Comment by swt2c** [comment]
@pmalhaire I didn't see the same error related to `kOSKextReturnSystemPolicy` when I tried compiling on the GMC.  However, I saw a different error when compiling mount_osxfuse that you (apparently) didn't see.  So I wonder if perhaps you are running an older version of the beta?

Anyway, I submitted https://github.com/osxfuse/support/pull/5 to fix the error with PostInstall and the error that I saw with building mount_osxfuse.

On the bridge support error, it seems to me that `gen_bridge_metadata` is completely broken on 10.13.  I submitted a bug report to Apple.  I found a hacky workaround, but I haven't found any way to work around the problem in osxfuse.

**Comment by pmalhaire** [comment]
@swt2c I have no update pending so I am at the last version of the os and Xcode

I had the same fix for post install which seems logical

what is you bridge workaround skipping it as I did or you got something better ?

**Comment by swt2c** [comment]
I'm running:
macOS 10.13 Beta (17A362a) (GM candidate - note this version doesn't show up as an upgrade in app store, you have to reinstall)
Xcode 9.0 (9A235) (official release, not beta)

The bridge workaround I used was to symlink /Applications/Xcode.app/Contents/Developer/Toolchains/OSX10.13.xctoolchain to /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain so that gen_bridge_metadata could find the libclang.dylib that it is looking for.

They haven't responded to the bug report yet.  They may not ever.

**Comment by pmalhaire** [comment]
Your damn right, beta to gm is not automated

**Comment by swt2c** [comment]
> Your damn right, beta to gm is not automated

Yep, it's annoying.  It looks as if the Xcode update might not be either, given that it doesn't seem to be offering you the final version of Xcode 9.  You might have to uninstall and reinstall that too.  :-(  Another 5GB download... :(

**Comment by pmalhaire** [comment]
@swt2c 
I updated Xcode and the `kOSKextReturnSystemPolicy` is a bug due to Xcode 9 beta, it's fixed with Xcode 9.
 
```bash
/Users/pierrot/osxfuse/support/load_osxfuse/fuse_kext.c:322:27: error: use of undeclared identifier 'kOSKextReturnSystemPolicy'
        } else if (ret == kOSKextReturnSystemPolicy) {
                          ^
1 error generated.
```

So I guess the other error you got `error when compiling mount_osxfuse`seems to be GMC regression. 
Hopefully the release will fix it, let wait the 25th

**Comment by swt2c** [comment]
@pmalhaire Benjamin merged my PR which fixes the mount_osxfuse compile problem.

I believe the only problem remaining now is the `gen_bridge_metadata` problem, for which I still don't see a way for osxfuse to work around.

**Comment by bfleischer** [comment]
> I believe the only problem remaining now is the gen_bridge_metadata problem, for which I still don't see a way for osxfuse to work around.

3.7.1 should compile with Xcode 9 after symlinking the toolchain, at least it worked for me. Let's wait and see if Apple addresses the toolchain issue. I don't think it will take long before Apple releases Xcode 9.0.1. This is not the only bug in Xcode 9.0.

@pmalhaire Thanks for reporting the issue. If I remember correctly `kOSKextReturnSystemPolicy` has been introduced in Xcode 9 beta 3. Maybe you were using an earlier beta.

**Comment by swt2c** [comment]
You are right - they are usually pretty quick to issue Xcode updates.

However, I think the problem with gen_bridge_metadata is in macOS itself (actually the problem seems to be with `/System/Library/BridgeSupport/ruby-2.3/bridgesupportparser.bundle`) as gen_bridge_metadata seems to be present even without installing Xcode or command line tools.  I have to wonder if bridgesupportparser.bundle shouldn't be linked against libclang.dylib (which doesn't exist without Xcode installed) in the first place, as the 10.12 version isn't.

**Comment by pmalhaire** [comment]
I tested again with official 10.13 release

the only issue left is [gen_bridge_metadata](https://github.com/osxfuse/osxfuse/issues/416#issuecomment-331225058) 

@bfleischer your choice to keep the issue or close it as we don't know when Apple will wake up.

**Comment by amirrajan** [comment]
Re: `gen_bridge_metadata` /cc @swt2c : https://gist.github.com/amirrajan/ead10b1acf6b60678e1b8332d571aae8

**Comment by swt2c** [comment]
@amirrajan Thanks, but I think the problem here is as I noted in an earlier comment - `gen_bridge_metadata` imports `bridgesupportparser` but `bridgesupportparser` can't resolve its shared library dependencies.    So the problem is really with `bridgesupportparser.`

**Comment by amirrajan** [comment]
@swt2c sorry, some clarification:

1. The [`high-sierra`](https://github.com/amirrajan/BridgeSupport/tree/high-sierra) branch is what you'd want to look at.

2. We had to patch [`gen_bridge_metadata`](https://github.com/amirrajan/BridgeSupport/blob/high-sierra/gen_bridge_metadata.rb) to keep it working for 10.12+ (subsequent patches were needed given that 10.13 now uses ruby 2.3).

3. This is "[version 2.0](https://github.com/amirrajan/BridgeSupport/tree/high-sierra/Versions/2.0/BridgeSupport) of BridgeSupport" built against 10.13 and ruby 2.3 (using this version removed the need for the symlink in Xcode).

**Comment by therealjimwolff** [comment]
I had a clean Xcode 9.1 and High Sierra 10.13.1. I couldn't build the high sierra branch of BridgeSupport either from github. But I was able to manually copy the stuff from "version 2.0" into the various places. Then reboot, and check the mod dates that the manual install of everything stuck (which I had to redo a few times and fresh reboot). Then I could build with these patches referenced above (Oct 7) comments.
